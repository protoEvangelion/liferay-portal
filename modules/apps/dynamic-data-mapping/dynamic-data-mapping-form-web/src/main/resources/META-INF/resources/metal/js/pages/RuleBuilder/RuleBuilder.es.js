import {Config} from 'metal-state';
import dom from 'metal-dom';
import {EventHandler} from 'metal-events';
import Component from 'metal-jsx';
import RuleList from '../../components/RuleList/index.es';
import RuleEditor from '../../components/RuleEditor/index.es';

/**
 * Builder.
 * @extends Component
 */

class RuleBuilder extends Component {
	static PROPS = {
		pages: Config.array().required(),

		rules: Config.arrayOf(
			Config.shapeOf(
				{
					actions: Config.arrayOf(
						Config.shapeOf(
							{
								action: Config.string(),
								label: Config.string(),
								target: Config.string()
							}
						)
					),
					conditions: Config.arrayOf(
						Config.shapeOf(
							{
								operands: Config.arrayOf(
									Config.shapeOf(
										{
											label: Config.string(),
											repeatable: Config.bool(),
											type: Config.string(),
											value: Config.string()
										}
									)
								),
								operator: Config.string()
							}
						)
					),
					logicalOperator: Config.string()
				}
			)
		),

		/**
		 * The path to the SVG spritemap file containing the icons.
		 * @default undefined
		 * @instance
		 * @memberof Form
		 * @type {!string}
		 */

		spritemap: Config.string().required()
	}

	static STATE = {

		/**
		 * @default
		 * @instance
		 * @memberof RuleBuilder
		 *
		 */

		mode: Config.oneOf(['view', 'edit']).value('view')
	};

	/**
	 * Continues the propagation of event.
	 * @param {!Event} event
	 * @private
	 */

	created() {
		this._eventHandler = new EventHandler();
	}

	/**
	 * Continues the propagation of event.
	 * @param {!Event} event
	 * @private
	 */

	_showRuleEdition() {
		this.setState(
			{
				mode: 'edit'
			}
		);
	}

	/**
	 * Show the rule screen to edit an existing rule. For now, this method does not receive the rule data for edition.
	 * @param {!Event} event
	 * @private
	 */

	_handleEditRuleClicked() {
		this._showRuleEdition();
	}

	/**
	 * Show the rule screen to create a new rule
	 * @param {!Event} event
	 * @private
	 */

	_handleAddRuleClick(event) {
		this._showRuleEdition();

		this._hideAddRuleButton(event.delegateTarget);
	}

	/**
	 * Continues the propagation of event.
	 * @param {!Event} event
	 * @private
	 */

	_hideAddRuleButton(element) {
		dom.addClasses(element, 'hide');
	}

	/**
	 * Continues the propagation of event.
	 * @param {!Event} event
	 * @private
	 */

	attached() {
		this._eventHandler.add(
			dom.on('#addFieldButton', 'click', this._handleAddRuleClick.bind(this)),
			dom.on('.rule-card-edit', 'click', this._handleEditRuleClicked.bind(this))
		);
	}

	/**
	 * Continues the propagation of event.
	 * @param {!Event} event
	 * @private
	 */

	dispose() {
		this._eventHandler.removeAllListeners();
	}

	/**
	 * Continues the propagation of event.
	 * @param {!Event} event
	 * @private
	 */

	render() {
		const {spritemap} = this.props;

		let ruleScreen;

		if (this.state.mode === 'edit') {
			ruleScreen = <RuleEditor />;
		}
		else {
			ruleScreen = <RuleList pages={this.props.pages} rules={this.props.rules} spritemap={spritemap} />;
		}

		return (
			<div class="container">
				{ruleScreen}
			</div>
		);
	}
}

export default RuleBuilder;
export {RuleBuilder};