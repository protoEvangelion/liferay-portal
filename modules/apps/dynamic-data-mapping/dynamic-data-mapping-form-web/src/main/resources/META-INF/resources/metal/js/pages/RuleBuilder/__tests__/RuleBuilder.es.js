import RuleBuilder from '../RuleBuilder.es';
import {dom as MetalTestUtil} from 'metal-dom';

const spritemap = 'icons.svg';
let component;

const configView = {
	mode: 'view',
	pages: [
		{
			rows: [
				{
					columns: [
						{
							fields: [
								{
									fieldName: 'text1',
									label: 'label text 1'
								},
								{
									fieldName: 'text2',
									label: 'label text 2'
								}
							]
						}
					]
				}
			]
		}
	],
	rules: [
		{
			actions: [
				{
					action: 'require',
					expression: '[x+2]',
					target: 'text1'
				}
			],
			conditions: [
				{
					operands: [
						{
							type: 'field',
							value: 'text1'
						},
						{
							type: 'field',
							value: 'text2'
						}
					],
					operator: 'equals-to'
				}
			],
			['logical-operator']: 'OR'
		}
	],
	spritemap
};

describe(
	'RuleBuilder',
	() => {
		beforeEach(
			() => {
				jest.useFakeTimers();
			}
		);
		afterEach(
			() => {
				if (component) {
					component.dispose();
				}
			}
		);
		it(
			'should render the list of rules when mode is set to view',
			() => {
				component = new RuleBuilder(configView);
				component.setState({mode: 'view'});
				expect(component).toMatchSnapshot();
			}
		);

		it(
			'should render rule screen editor when click edit button',
			() => {
				MetalTestUtil.enterDocument('<button id="addFieldButton" class="hide"></button>');

				const ruleCardEdit = document.querySelector('.rule-card-edit');

				component = new RuleBuilder(configView);

				MetalTestUtil.triggerEvent(ruleCardEdit, 'click', {});

				jest.runAllTimers();

				expect(component).toMatchSnapshot();
			}
		);

		it(
			'should render rule screen editor when click add rule button',
			() => {
				MetalTestUtil.enterDocument('<button id="addFieldButton" class="hide"></button>');

				const ruleCardEdit = document.querySelector('#addFieldButton');

				component = new RuleBuilder(configView);

				MetalTestUtil.triggerEvent(ruleCardEdit, 'click', {});

				jest.runAllTimers();

				expect(component).toMatchSnapshot();
			}
		);
	}
);