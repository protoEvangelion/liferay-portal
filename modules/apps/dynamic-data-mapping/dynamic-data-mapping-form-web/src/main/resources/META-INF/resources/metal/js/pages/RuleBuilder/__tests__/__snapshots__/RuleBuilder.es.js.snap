// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuleBuilder should render rule screen editor when click add rule button 1`] = `
<div class="container">
  <div class="container"></div>
</div>
`;

exports[`RuleBuilder should render rule screen editor when click edit button 1`] = `
<div class="container">
  <div class="container"></div>
</div>
`;

exports[`RuleBuilder should render the list of rules when mode is set to view 1`] = `
<div class="container">
  <div class="form-builder-rule-builder-container">
    <h1 class="form-builder-section-title text-default">Rule Builder</h1>
    <div class="liferay-ddm-form-rule-rules-list-container">
      <div class="form-builder-rule-list">
        <ul class="ddm-form-body-content form-builder-rule-builder-rules-list tabular-list-group">
          <li class="list-group-item">
            <div class="clamp-horizontal list-group-item-content">
              <p class="form-builder-rule-builder-rule-description text-default">
                <b>if </b>
                <span>field </span>
                <span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1">
                  <span class="text-truncate-inline">
                    <span class="text-truncate">label text 1</span>
                  </span>
                </span>
                <b class="text-lowercase">
                  <em> is-equal-to </em>
                </b>
                <span>field </span>
                <span class="label label-lg label-secondary" data-original-title="label text 2" title="label text 2">
                  <span class="text-truncate-inline">
                    <span class="text-truncate">label text 2</span>
                  </span>
                </span>
                <br>
                <span>
                  <b>require </b>
                </span>
                <span class="label label-lg label-secondary" data-original-title="label text 1" title="label text 1">
                  <span class="text-truncate-inline">
                    <span class="text-truncate">label text 1</span>
                  </span>
                </span>
              </p>
            </div>
            <div class="list-group-item-field">
              <div class="card-col-field">
                <div class="dropdown dropdown-action">
                  <ul class="dropdown-menu dropdown-menu-right">
                    <li class="rule-card-edit" data-card-id="0">
                      <a href="javascript:;">edit</a>
                    </li>
                    <li class="rule-card-delete" data-card-id="0">
                      <a href="javascript:;">delete</a>
                    </li>
                  </ul>
                  <a class="component-action dropdown-toggle" data-toggle="dropdown" href="#1"></a>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
